### PR TITLE
Addition of Cisco 8201-32FH device-type router model

### PR DESCRIPTION
**Pull request by tonnerrek**
the model is based on ASR-9901; all the modifications were done based on experience with hardware, access to CLI and validated with official documentation.

### CODE REVIEW

File: device-types/Cisco/8201-32FH.yaml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+console-ports:
+  - name: con 0
+    type: rj-45
+power-ports:
+  - name: PM0
+    type: iec-60320-c14
+    maximum_draw: 2000
+  - name: PM1
+    type: iec-60320-c14
+    maximum_draw: 2000
```
Comment (DanSheps) [inline device-types/Cisco/8201-32FH.yaml:19]: Should be modules

File: device-types/Cisco/8201-32FH.yaml [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+interfaces:
+  - name: Mg0/RP0/CPU0/0
+    type: 1000base-t
```
Comment (DanSheps) [inline device-types/Cisco/8201-32FH.yaml:22]: Is this the actual name?

**Review by DanSheps — CHANGES_REQUESTED** [review]

File: device-types/Cisco/8201-32FH.yaml (hunk 2 above)
Comment (DanSheps) [reply device-types/Cisco/8201-32FH.yaml:22]: Looks like it should be:  MgmtEth0/RSP0RP0/CPU0/0

There may also be a:

MgmtEth0/RSP0RP0/CPU0/1
MgmtEth0/RSP1RP1/CPU0/0
MgmtEth0/RSP1RP1/CPU0/1

**Comment by tonnerrek** [comment]
I'm a network engineer and Netbox enthusiast, but I'm new to git,  and especially new to github forking/review model. I'm sorry in advance for slow responsiveness and some clumsiness in git. 
I plan to add many more device models, and possibly modules types. I'm happy to share them with the community.

**Comment by danner26** [comment]
Do you still need this? @tonnerrek

**Comment by tonnerrek** [comment]
Hi @danner26 , I'm sorry for late reply - I had some days off.

I think I've implemented the changes requested by @DanSheps, and the model should be 100% fine now following your standards and Cisco's naming convention. I use this model in lab setup at work, and I thought that thewhole Netbox  community could benefit from it.

**Comment by danner26** [comment]
@tonnerrek Can you please send us an export of the configration with the interface names? Some of them look a little weird. I havent used this specific model so they might be 100% correct, I just want to confirm. Thanks!

File: device-types/Cisco/ASR-9902.yaml [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+- name: FAN1
+  label: ''
+  position: '1'
+  description: Fan tray
+- name: FAN2
+  label: ''
+  position: '2'
+  description: Fan tray
+
+---
```
Comment (danner26) [inline device-types/Cisco/ASR-9902.yaml:40]: Please correct this file, it looks like you have two seperate files in one. Only one device should be defined per file

**Review by danner26 — CHANGES_REQUESTED** [review]

**Comment by tonnerrek** [comment]
> @tonnerrek Can you please send us an export of the configration with the interface names? Some of them look a little weird. I havent used this specific model so they might be 100% correct, I just want to confirm. Thanks!

Here you go:
`RP/0/RP0/CPU0:ipt-zhb921-g-cme-01#show run | inc ^interface
Thu Jan  5 15:31:33.966 CET
Building configuration...
interface Bundle-Ether919
interface Bundle-Ether1000
interface Bundle-Ether9192
interface Bundle-Ether9193
interface Loopback0
interface Loopback1
interface MgmtEth0/RP0/CPU0/0
interface FourHundredGigE0/0/0/0
interface FourHundredGigE0/0/0/1
interface FourHundredGigE0/0/0/2
interface FourHundredGigE0/0/0/3
interface FourHundredGigE0/0/0/4
interface FourHundredGigE0/0/0/5
interface FourHundredGigE0/0/0/6
interface FourHundredGigE0/0/0/7
interface FourHundredGigE0/0/0/8
interface FourHundredGigE0/0/0/9
interface FourHundredGigE0/0/0/10
interface FourHundredGigE0/0/0/11
interface FourHundredGigE0/0/0/12
interface FourHundredGigE0/0/0/13
interface FourHundredGigE0/0/0/14
interface FourHundredGigE0/0/0/15
interface FourHundredGigE0/0/0/16
interface FourHundredGigE0/0/0/17
interface FourHundredGigE0/0/0/18
interface FourHundredGigE0/0/0/19
interface FourHundredGigE0/0/0/20
interface HundredGigE0/0/0/21/0
interface HundredGigE0/0/0/21/1
interface HundredGigE0/0/0/21/2
interface HundredGigE0/0/0/21/3
interface FourHundredGigE0/0/0/22
interface FourHundredGigE0/0/0/23
interface FourHundredGigE0/0/0/24
interface FourHundredGigE0/0/0/25
interface FourHundredGigE0/0/0/26
interface FourHundredGigE0/0/0/27
interface FourHundredGigE0/0/0/28
interface FourHundredGigE0/0/0/29
interface FourHundredGigE0/0/0/30
interface FourHundredGigE0/0/0/31`

**Comment by danner26** [comment]
Thanks @tonnerrek ! I appreciate it
Can you please fix `device-types/Cisco/ASR-9902.yaml` so we can continue?

**Comment by danner26** [comment]
@tonnerrek for the ASR 9902 you start with adapter `GigabitEthernet0/0/0/1`, then goes to 7, and next starts at `GigabitEthernet0/0/0/34`
`HundredGigE0/0/0/20` is that start
then goes to `TenGigE0/0/0/8`

Is this configuration correct?

**Comment by tonnerrek** [comment]
@danner26 It's incorrect. I'm on it, but due to some ongoing meetings my progress is very slow. Please give me ca. 10 minutes and I'll let you know once I feel good with ASR-9902 model.

**Comment by danner26** [comment]
No rush at all. Why dont you tag me once you are done making your changes, then ill review again! Thank you for your efforts

**Comment by danner26** [comment]
Please also remove the following lines;
```
    mgmt_only: false
    label: ""
    description: ""
```

mgmt_only defaults to false, so no need to define it as false
label defaults to none, no need to define it as none
description defaults to none, no need to define it as none

Thanks! @tonnerrek

**Comment by danner26** [comment]
@tonnerrek for these:
```
- name: TenGigE0/0/0/24
    type: 25gbase-x-sfp28
    description: default 25GE (and names prefix is TF, but can be configured as 10GE)
```

If they default to 25GE, why dont we mark them as such. I think it would be desirable to have the device default, and if the user explicitly changes it on the device, then they would need to explicitly change it here as well.

**Comment by tonnerrek** [comment]
Hello @danner26,

I removed the "_flat" models, that were built to for legacy netbox versions. Such version is still running at my work, but I will keep them in a separate branch.
All the normal (non _flat) models should be composed of default values, e.g. N540-ACC-SYS, where the relevant ports are 25GbE out of the box.

**Comment by tonnerrek** [comment]
Hello @danner26,

I removed the "_flat" models, that were built to for legacy netbox versions. Such version is still running at my work, but I will keep them in a separate branch.
All the normal (non _flat) models should be composed of default values, e.g. N540-ACC-SYS, where the relevant ports are 25GbE out of the box.

**Comment by danner26** [comment]
Hey @tonnerrek can we just confirm, is this the correct order for the ASR-9902? You have `TenGigE0/0/0/7` then `HundredGigE0/0/0/8` and then once it gets to 12 it goes back to `TenGigE0/0/0/12`
I do not have one of these units to confirm, so if you could please confirm and/or send a copy of the names from the terminal that would be great!

**Comment by tonnerrek** [comment]
Hi @danner26 ,

On ASR9902 there are only built-in ports, but depending on port-group configuration either 100GbE or 10GbE ports are active. Currently, on my node I have 4x100GbE configured, therefore no TenGig interface is listed:

`RP/0/RP0/CPU0:ipt-lss923-b-crn-01#show int brief
Fri Jan  6 16:57:34.285 CET

               Intf       Intf        LineP              Encap  MTU        BW
               Name       State       State               Type (byte)    (Kbps)
--------------------------------------------------------------------------------
                Lo0          up          up           Loopback  1500          0
                Lo1          up          up           Loopback  1500          0
                Nu0          up          up               Null  1500          0
     Mg0/RP0/CPU0/0          up          up               ARPA  1514          0
     Mg0/RP1/CPU0/0          up          up               ARPA  1514    1000000
          Hu0/0/0/8  admin-down  admin-down               ARPA  1514  100000000
          Hu0/0/0/9  admin-down  admin-down               ARPA  1514  100000000
         Hu0/0/0/10          up          up               ARPA  9216  100000000
         Hu0/0/0/11  admin-down  admin-down               ARPA  1514  100000000
         Hu0/0/0/36          up          up               ARPA  9216  100000000
         Hu0/0/0/37  admin-down  admin-down               ARPA  1514  100000000
         Hu0/0/0/38  admin-down  admin-down               ARPA  1514  100000000
         Hu0/0/0/39  admin-down  admin-down               ARPA  1514  100000000`

`RP/0/RP0/CPU0:ipt-lss923-b-crn-01#show controllers np ports all
Fri Jan  6 16:56:09.547 CET

                Node: 0/0/CPU0:
----------------------------------------------------------------

NP Bridge Fia                       Ports
-- ------ --- ---------------------------------------------------
0  --     0   HundredGigE0/0/0/11, HundredGigE0/0/0/10, HundredGigE0/0/0/9, HundredGigE0/0/0/8
1  --     1   HundredGigE0/0/0/37, HundredGigE0/0/0/36, HundredGigE0/0/0/39, HundredGigE0/0/0/38`

I decided that it's good that all ports are listed, even though they all cannot be used at the same time.

Cisco states:
> 2 ports that support QSFP-DD-based 100GE pluggables
> 6 ports that support QSFP28-based 100GE pluggables
> 16 ports that support SFP28-based 25GE/10GE dual-rate pluggables
> 24 ports that support SFP+- based LAN/WAN (OTN) pluggables

so I could even specify that ports 0-7 and 40-47 are actually 25GbE (SFP28). What would you suggest? 
https://www.cisco.com/c/en/us/td/docs/iosxr/asr9000/hardware-install/9001-hig/b-asr9001-hardware-installation-guide/b-asr9001-hardware-installation-guide_chapter_01.html#:~:text=The%20router%20consists%20of%20the%20fixed%20board%20or%20line%20card%20with%2048%20ports.%20The%2048%20ports%20are%20grouped%20in%20slice%200%20and%20slice%201.%20The%20slice%200%20is%20color%20coded%20in%20blue%2C%20and%20slice%201%20in%20purple.%20All%20ports%20support%20MACSec%3A

**Comment by danner26** [comment]
Ah I see, okay lets stick with what I said previously. **Whatever comes default on the device should be what we have listed**
Once you confirm the code currently committed is the device defaults, then we can merge this. Thanks for all your work (and I appreciate you teaching me something)!

**Comment by tonnerrek** [comment]
Here are the default port speed values:

```
RP/0/RP0/CPU0:ipt-lss923-b-crn-01#show controllers np valid-port-groups all location 0/0/CPU0
Fri Jan  6 19:49:52.631 CET

                Node: 0/0/CPU0:
----------------------------------------------------------------

 NP0: Valid Port Groups
----+---------------------+---------------------+---------------------+---------------------
SEL | Group 0             | Group 1             | Group 2             | Group 3
----+---------------------+---------------------+---------------------+---------------------
DFLT 1x100GE               1x100GE               10x10GE               10x10GE
1    1x100GE               1x100GE               10x10GE/5x1GE_5x10GE  10x10GE/5x1GE_5x10GE
2    1x100GE               1x100GE               4x25GE                10x10GE/5x1GE_5x10GE
3    1x100GE               4x25GE                4x25GE                1x100GE
4    1x100GE               1x100GE               1x100GE               1x100GE
Notes:
  DFLT: Port 11(QSFP-DD), Port 10(QSFP28),    Port 0-7(SFP10g L),22-23(SFP10g L/O),                              Port 12-21(SFP10g L/O) (default)
     1: Port 11(QSFP-DD), Port 10(QSFP28),    Port 1/3/5/7/23(SFP1g),Port 0/2/4/6(SFP10g L),Port 22(SFP10g L/O), Port 13/15/17/19/21(SFP1g),Port 12/14/16/18/20(SFP10g L/O)
     2: Port 11(QSFP-DD), Port 10(QSFP28),    Port 0-7(SFP10g L),22-23(SFP10g L/O),                              Port 13/15/17/19/21(SFP1g),Port 12/14/16/18/20(SFP10g L/O)
     3: Port 11(QSFP-DD), Port 10(QSFP28),    Port 1/3/5/7/23(SFP1g),Port 0/2/4/6(SFP10g L),Port 22(SFP10g L/O), Port 12-21(SFP10g L/O)
     4: Port 11(QSFP-DD), Port 10(QSFP28),    Port 4-7(SFP25g L),                                                Port 12-21(SFP10g L/O)
     5: Port 11(QSFP-DD), Port 10(QSFP28),    Port 4-7(SFP25g L),                                                Port 13/15/17/19/21(SFP1g),Port 12/14/16/18/20(SFP10g L/O)
     6: Port 11(QSFP-DD), Port 0-3(SFP25g L), Port 4-7(SFP25g L),                                                Port 8(QSFP28)
     7: Port 11(QSFP-DD), Port 10(QSFP28),    Port 9(QSFP28),                                                    Port 8(QSFP28)

 NP1: Valid Port Groups
----+---------------------+---------------------+---------------------+---------------------
SEL | Group 0             | Group 1             | Group 2             | Group 3
----+---------------------+---------------------+---------------------+---------------------
DFLT 1x100GE               1x100GE               10x10GE               10x10GE
1    1x100GE               1x100GE               10x10GE/5x1GE_5x10GE  10x10GE/5x1GE_5x10GE
2    1x100GE               1x100GE               4x25GE                10x10GE/5x1GE_5x10GE
3    1x100GE               4x25GE                4x25GE                1x100GE
4    1x100GE               1x100GE               1x100GE               1x100GE
Notes:
  DFLT: Port 37(QSFP-DD), Port 36(QSFP28),      Port 40-47(SFP10g L),24-25(SFP10g L/O),                                    Port 26-35(SFP10g L/O) (default)
     1: Port 37(QSFP-DD), Port 36(QSFP28),      Port 25/41/43/45/47(SFP1g),Port 40/42/44/46(SFP10g L),Port 24(SFP10g L/O), Port 27/29/31/33/35(SFP1g),Port 26/28/30/32/34(SFP10g L/O)
     2: Port 37(QSFP-DD), Port 36(QSFP28),      Port 40-47(SFP10g L),24-25(SFP10g L/O),                                    Port 27/29/31/33/35(SFP1g),Port 26/28/30/32/34(SFP10g L/O)
     3: Port 37(QSFP-DD), Port 36(QSFP28),      Port 25/41/43/45/47(SFP1g),Port 40/42/44/46(SFP10g L),Port 24(SFP10g L/O), Port 26-35(SFP10g L/O)
     4: Port 37(QSFP-DD), Port 36(QSFP28),      Port 40-43(SFP25g L),                                                      Port 26-35(SFP10g L/O)
     5: Port 37(QSFP-DD), Port 36(QSFP28),      Port 40-43(SFP25g L),                                                      Port 27/29/31/33/35(SFP1g),Port 26/28/30/32/34(SFP10g L/O)
     6: Port 37(QSFP-DD), Port 44-47(SFP25g L), Port 40-43(SFP25g L),                                                      Port 38(QSFP28)
     7: Port 37(QSFP-DD), Port 36(QSFP28),      Port 39(QSFP28),                                                           Port 38(QSFP28)
     
```

It's a mix of 100GbE and 10GbE ports. I'll reflect this state in the model and let you once it's done (most likely tomorrow).

**Comment by danner26** [comment]
Much appreciated! Thank you @tonnerrek

**Comment by danner26** [comment]
@tonnerrek question for you, on `device-types/Cisco/ASR-9903.yaml` you have module bay positions that share the same number. For instance there are more than 1 position `0` and position `1`, etc. Is there any chance the modules plugged into these would share the same interface names? If so, that will cause a clash in NetBox.

**Comment by tonnerrek** [comment]
@danner26 you're right, thanks for spotting this!
I must admit that I did not think it though well. My assumption was that I have various module-bay slot types (e.g. FAN, PSU, line-card, RSP), but indeed it is not enforced, so I could insert a line-card into FAN-designated bay. 

Should I renumber the module bays? or there is another way to mitigate such position overlap? Any advice from you is more than appreciated!

**Comment by danner26** [comment]
@tonnerrek if you check out one of [my recent commits](https://github.com/netbox-community/devicetype-library/commit/fbae035870cd7dc656fd8b3ecad05c8000ed04f6) I laid the module bays out such that the name was used in the interface name how I wanted.

For instance, the PSUs needed to be listed as `PSU1` and `PSU2`, so I named the bays 1 and 2, so that when you plug a PSU module in the name will be `PSU{module}` which will evaluate to either `PSU1` or `PSU2`. Whereas the modular controllers need to be listed as `Controller A` and `Controller B`, because that is how Dell lays them out in the config. Maybe you can follow a similar model?

**Comment by danner26** [comment]
@tonnerrek actually it will be `mgmtport_a` and `mgmtport_b`. The `{module}` definition will use the `name: ` part

**Comment by tonnerrek** [comment]
@danner26 Did the behavior change in some recent releases? I'm running 3.3.5 and I just instantiated this Dell storage array:
image
On my side the ports inherited from modules get "position:" value

**Comment by danner26** [comment]
Hm interesting, I must have misinterpreted that then when I was reading over the docs. I was working on that as I was rushing out the door

So anyway I stand corrected

**Comment by danner26** [comment]
Since position is a string, all we will have to do is override that to a and b for the position. Thanks for catching that!

**Comment by tonnerrek** [comment]
@danner26 I fixed module-bay position numbering, now it should be all good :-)

**Review by danner26 — APPROVED** [review]

**Comment by danner26** [comment]
Thanks for your diligence on this @tonnerrek! We really appreciate it. Your changes will be merged

**Comment by tonnerrek** [comment]
Thank you very much @danner26 for your help and patience! I'm a rookie in git, and any kind of software development. ;-)
Now having a very valuable exechange with you, the next time I create the pull-request the quality of my "code" will be significantly better.

Cheers

**Pull request merged**